### PR TITLE
Display errors in cells when we run into kernel issues

### LINKED ISSUE #9269: Restore error message when not installing IPYKernel or Jupyter Dependencies
The error message displayed when users cancel the prompt to install ipykernel has changed (the new error is as per screenshot).

Previously we'd display a message indicating the fact that the cell could not be run as ipykernle is not installed.
Yes, the user hit cancel here, but they don't know what ipykernle prompt is, and we chose to display the detailed error message in the output as well (asking with instructions to unblock themselves)

I think we should bring back the old error message.

image

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #9269

### CODE REVIEW

File: src/client/common/errors/types.ts [hunk 1]
```diff
@@ -10,6 +11,17 @@ export abstract class BaseError extends Error {
     }
 }
 
+export abstract class BaseKernelError extends Error {
```
Comment (DonJayamanne) [inline src/client/common/errors/types.ts:14]: All kernel errors will have the `kernelConnectionMetadata`
Ideally we should include the `resource`, but don't need that just yet, hence not adding it now.

File: src/client/datascience/errors/errorHandler.ts [hunk 2]
```diff
@@ -135,11 +164,16 @@ export class DataScienceErrorHandler implements IDataScienceErrorHandler {
             );
             if (failureInfo) {
                 void this.showMessageWithMoreInfo(failureInfo?.message, failureInfo?.moreInfoLink);
+            } else if (err instanceof BaseError) {
+                const message = getCombinedErrorMessage(getErrorMessageFromPythonTraceback(err.stdErr) || err.stdErr);
+                void this.showMessageWithMoreInfo(message);
+            } else {
+                void this.showMessageWithMoreInfo(err.message);
```
Comment (DonJayamanne) [inline src/client/datascience/errors/errorHandler.ts:171]: Always display errors regardless of the error.

File: src/client/datascience/errors/kernelPortNotUsedTimeoutError.ts [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
-    constructor(public readonly kernelConnection: KernelConnectionMetadata) {
+export class KernelPortNotUsedTimeoutError extends BaseKernelError {
+    constructor(kernelConnection: KernelConnectionMetadata) {
         super(
             'timeout',
             DataScience.rawKernelStartFailedDueToTimeout().format(
                 getDisplayNameOrNameOfKernelConnection(kernelConnection)
-            )
+            ),
+            kernelConnection
```
Comment (DonJayamanne) [inline src/client/datascience/errors/kernelPortNotUsedTimeoutError.ts:16]: All kernel errors will have `kernelConnection` so we have more context.

File: src/client/datascience/jupyter/kernels/helpers.ts [hunk 4]
```diff
@@ -1950,10 +1949,6 @@ export async function handleKernelError(
 
     switch (handleResult) {
         case KernelInterpreterDependencyResponse.cancel:
-            throw new CancellationError(
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/helpers.ts:1953]: rethrow original error

File: src/client/datascience/kernel-launcher/kernelProcess.ts [hunk 5]
```diff
@@ -191,29 +190,7 @@ export class KernelProcess implements IKernelProcess {
                     return;
                 }
                 traceError('Kernel died', error, stderr);
-                if (error instanceof PythonKernelDiedError) {
```
Comment (DonJayamanne) [inline src/client/datascience/kernel-launcher/kernelProcess.ts:194]: This error is never thrown anywhere, it was only thrown in the case of kernels being launched using daemon (which isn't used anymore)

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/jupyter/kernels/helpers.ts (hunk 4 above)
Comment (rchiodo) [reply src/client/datascience/jupyter/kernels/helpers.ts:1953]: I imagine this fixed the main problem.

File: src/client/datascience/notebook/vscodeNotebookController.ts [hunk 6]
```diff
@@ -385,8 +386,11 @@ export class VSCodeNotebookController implements Disposable {
             }
             return await kernel.executeCell(cell);
         } catch (ex) {
+            const errorHandler = this.serviceContainer.get<IDataScienceErrorHandler>(IDataScienceErrorHandler);
+
             // If there was a failure connecting or executing the kernel, stick it in this cell
-            return displayErrorsInCell(cell, execution, ex.toString());
+            displayErrorsInCell(cell, execution, await errorHandler.getErrorMessageForDisplayInCell(ex));
+            return NotebookCellExecutionState.Idle;
```
Comment (rchiodo) [inline src/client/datascience/notebook/vscodeNotebookController.ts:393]: Wouldn't this be an error? Idle seems wrong?

Comment (DonJayamanne) [reply src/client/datascience/notebook/vscodeNotebookController.ts:393]: This is the execution state, not about the state after the execution , i.e. not execution result.
Valid values from VS Code are running, not running or Pending.

File: src/client/datascience/jupyter/kernels/helpers.ts (hunk 4 above)
Comment (DonJayamanne) [reply src/client/datascience/jupyter/kernels/helpers.ts:1958]: Yes

**Pull request merged**